crates/services/txpool_v2/src/tests/pool.rs

```diff
@@ -0,0 +1,1627 @@
+    let mut context = PoolContext::default();
+
+    let (_, coin_input) = context.setup_coin();
+
+    let tx1 = TransactionBuilder::script(vec![], vec![])
+        .tip(10)
+        .max_fee_limit(10)
+        .script_gas_limit(GAS_LIMIT)
+        .add_input(coin_input.clone())
+        .finalize_as_transaction();
+
+    let tx2 = TransactionBuilder::script(vec![], vec![])
+        .tip(20)
+        .max_fee_limit(20)
+        .script_gas_limit(GAS_LIMIT)
+        .add_input(coin_input)
+        .finalize_as_transaction();
+
+    let tx1_id = tx1.id(&ChainId::default());
+    let mut txpool = context.build();
+    let tx1 = check_unwrap_tx(tx1, &txpool.config).await;
+    let tx2 = check_unwrap_tx(tx2, &txpool.config).await;
+
+    let results = txpool
+        .insert(vec![
+            check_tx_to_pool(tx1.clone()),
+            check_tx_to_pool(tx2.clone()),
+        ])
+        .unwrap();
+    assert_eq!(results.len(), 2);
+    assert!(results[0].is_ok());
+    let vec = results[1].as_ref().unwrap();
+    assert_eq!(vec[0].id(), tx1_id, "Tx1 id should be removed");
+}
+
+#[tokio::test]
+async fn underpriced_tx1_not_included_coin_collision() {
+    let mut context = PoolContext::default();
+
+    let (_, gas_coin) = context.setup_coin();
+    let (output, unset_input) = context.create_output_and_input(20);
+    let tx1 = TransactionBuilder::script(vec![], vec![])
+        .tip(20)
+        .max_fee_limit(20)
+        .script_gas_limit(GAS_LIMIT)
+        .add_input(gas_coin)
+        .add_output(output)
+        .finalize_as_transaction();
+
+    let input = unset_input.into_input(UtxoId::new(tx1.id(&Default::default()), 0));
+
+    let tx2 = TransactionBuilder::script(vec![], vec![])
+        .tip(20)
+        .max_fee_limit(20)
+        .script_gas_limit(GAS_LIMIT)
+        .add_input(input.clone())
+        .finalize_as_transaction();
+
+    let tx3 = TransactionBuilder::script(vec![], vec![])
+        .tip(10)
+        .max_fee_limit(10)
+        .script_gas_limit(GAS_LIMIT)
+        .add_input(input)
+        .finalize_as_transaction();
+
+    let mut txpool = context.build();
+    let tx1_checked = check_unwrap_tx(tx1.clone(), &txpool.config).await;
+    let tx2_checked = check_unwrap_tx(tx2.clone(), &txpool.config).await;
+    let tx3_checked = check_unwrap_tx(tx3, &txpool.config).await;
+
+    let results = txpool
+        .insert(vec![
+            check_tx_to_pool(tx1_checked),
+            check_tx_to_pool(tx2_checked),
+            check_tx_to_pool(tx3_checked),
+        ])
+        .unwrap();
+    assert_eq!(results.len(), 3);
+    assert!(results[0].is_ok());
+    assert!(results[1].is_ok());
+    let err = results[2].as_ref().expect_err("Tx3 should be Err, got Ok");
+
+    assert!(matches!(err, Error::Collided(_)));
+}
+
+// TODO: Why this should error ?
+// #[tokio::test]
+// async fn overpriced_tx_contract_input_not_inserted() {
+//     let mut context = PoolContext::default();
```

Comment:
Alright, if I'm reading this correctly we should not allow transactions that take contract inputs, with a higher tip then the contract has set.

It's not obvious to me why this is the desired behavior, but my guess is that it is to prevent someone from draining contract funds as tips somehow. I need to improve my understanding on how contract inputs work in general 😅.

In either case, adding this validation sounds like a natural follow-up. Can you create a ticket and reference it here in the todo comment?

Reply:
I placed this comment because I didn'tt understood this logic. After discussion with Green this test is not relevant anymore and contract creation are dependencies of contract usage.

Removed the test and the todo